planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/ReadablePmtiles.java

```diff
@@ -0,0 +1,161 @@
+package com.onthegomap.planetiler.pmtiles;
+
+import com.onthegomap.planetiler.archive.ReadableTileArchive;
+import com.onthegomap.planetiler.geo.TileCoord;
+import com.onthegomap.planetiler.util.CloseableIterator;
+import com.onthegomap.planetiler.util.Gzip;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class ReadablePmtiles implements ReadableTileArchive {
+  private SeekableByteChannel channel;
+
+  public ReadablePmtiles(SeekableByteChannel channel) {
+    this.channel = channel;
+  }
+
+  public static WriteablePmtiles.Entry findTile(List<WriteablePmtiles.Entry> entries, long tileId) {
+    int m = 0;
+    int n = entries.size() - 1;
+    while (m <= n) {
+      int k = (n + m) >> 1;
+      long cmp = tileId - entries.get(k).tileId();
+      if (cmp > 0) {
+        m = k + 1;
+      } else if (cmp < 0) {
+        n = k - 1;
+      } else {
+        return entries.get(k);
+      }
+    }
+
+    if (n >= 0) {
+      if (entries.get(n).runLength() == 0) {
+        return entries.get(n);
+      }
+      if (tileId - entries.get(n).tileId() < entries.get(n).runLength()) {
+        return entries.get(n);
+      }
+    }
+    return null;
+  }
+
+  // this is very inefficient
+  @Override
+  public byte[] getTile(int x, int y, int z) {
+    try {
+      var tileId = TileCoord.ofXYZ(x, y, z).hilbertEncoded();
+      var header = getHeader();
+
+      long dirOffset = header.rootDirOffset();
+      long dirLength = header.rootDirLength();
+
+      for (int depth = 0; depth <= 3; depth++) {
+        this.channel.position(dirOffset);
+        var buf = ByteBuffer.allocate((int) dirLength);
+        this.channel.read(buf);
+
+        byte[] dirBytes = buf.array();
+        if (header.internalCompression() == WriteablePmtiles.Compression.GZIP) {
+          dirBytes = Gzip.gunzip(dirBytes);
+        }
+
+        var dir = WriteablePmtiles.deserializeDirectory(dirBytes);
+        var entry = findTile(dir, tileId);
+        if (entry != null) {
+          if (entry.runLength() > 0) {
+            var buf2 = ByteBuffer.allocate(entry.length());
+            this.channel.position(header.tileDataOffset() + entry.offset());
+            this.channel.read(buf2);
+            return buf2.array();
+          } else {
+            dirOffset = header.leafDirectoriesOffset() + entry.offset();
+            dirLength = entry.length();
+          }
+        } else {
+          return null;
+        }
+
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException("Could not get tile", e);
+    }
+
+    return null;
+  }
+
+  public WriteablePmtiles.Header getHeader() throws IOException {
```

Comment:
I left most of the general purpose classes like `Header` in `WriteablePmtiles`, do we need to extract all those out into a Pmtiles.Util` package?

Reply:
created new `Pmtiles` common class